lms/resources/_js_config.py

```diff
@@ -14,28 +14,52 @@ def __init__(self, context, request):
         self._context = context
         self._request = request
 
+        # A dict of URLs for the frontend to use.
+        self._urls = {}
+
     @property
     @functools.lru_cache()
     def config(self):
         """
         Return the configuration for the app's JavaScript code.
 
-        This is a mutable config dict. It can be accessed, for example by
-        views, and they can mutate it or add their own view-specific config
-        settings. The modified config object will then be passed to the
-        JavaScript code in the response page.
+        This is a cached property (with @functools.lru_cache()) so that it's
+        mutable. Code outside of this method can do self.config["foo"] = "bar"
+        and, because self.config always returns the same cached dict and not a
+        new dict every time, the mutation will be preserved.
+
+        This is implemented as a property rather than having the __init__()
+        method initialize the self.config dict as a simple attribute, because
+        generating the dict can raise.  If it's going to raise we don't want
+        that to happen until as late as possible (not until someone actually
+        tries to access the dict). If JSConfig.__init__() were to raise that
+        would change the location of the crash in Pyramid's request processing
+        pipeline and could change the type of error response as a result.
 
         :rtype: dict
         """
         return {
-            "authToken": self._auth_token,
-            "debug": self._debug,
-            "hypothesisClient": self._hypothesis_config,
-            "rpcServer": self._rpc_server_config,
+            # The auth token that the JavaScript code will use to authenticate
+            # itself to our own backend's APIs.
+            "authToken": self._auth_token(),
+            # Some debug information, currently used in the Gherkin tests.
+            "debug": self._debug(),
+            # The config object for the Hypothesis client.
+            # Our JSON-RPC server passes this to the Hypothesis client over
+            # postMessage.
+            "hypothesisClient": self._hypothesis_client,
+            # The config object for our JSON-RPC server.
+            "rpcServer": {
+                "allowedOrigins": self._request.registry.settings[
+                    "rpc_allowed_origins"
+                ],
+            },
+            # A dict of URLs for the frontend to use.
+            # For example: API endpoints for the frontend to call would go in
+            # here.
             "urls": self._urls,
```

Comment:
If `rpcServer` above can be inlined I think the `_urls` dict can be do. Unless I'm missing something about why this is in the constructor?

Reply:
It's because `urls` will get mutated in future whereas `rpcServer` won't